### PR TITLE
feat!: rejects invalid empty range proofs

### PR DESCRIPTION
**Pull request by staheri14**
## Overview
Resolves https://github.com/celestiaorg/nmt/issues/164 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: proof_test.go [hunk 1]
```diff
@@ -13,6 +13,76 @@ import (
...
 
+// TestVerifyNamespace_EmptyProof tests the correct behaviour of VerifyNamespace for valid and invalid empty proofs.
+func TestVerifyNamespace_EmptyProof(t *testing.T) {
+	// create a tree with 4 leaves
+	nIDSize := 1
+	tree := exampleNMT(nIDSize, 1, 2, 3, 4)
+	root, err := tree.Root()
+	require.NoError(t, err)
+
+	// build a proof for an NID that is outside tree range of the tree
```
Comment (rootulp) [inline proof_test.go:24]: "outside tree range of the tree" seems slightly wrong but I'm not sure what it meant to say. Perhaps

```suggestion
	// build a proof for an NID that is outside the range of the tree
```

File: proof_test.go [hunk 2]
```diff
@@ -13,6 +13,76 @@ import (
...
+	// start = end = 0, nodes = empty
+	nID0 := []byte{0}
+	validEmptyProofZeroRange, err := tree.ProveNamespace(nID0)
+	require.NoError(t, err)
+
+	// build a proof for an NID that is outside tree range of the tree
```
Comment (rootulp) [inline proof_test.go:30]: same

**Review by rootulp — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]
:+1: :+1: LGTM

File: proof_test.go (hunk 1 above)
Comment (staheri14) [reply proof_test.go]: Thanks, addressed https://github.com/celestiaorg/nmt/pull/180/commits/8a276366b706a678c296a10aa718fc1d0667fb14

**Review by evan-forbes — DISMISSED** [review]

File: proof.go [hunk 3]
```diff
@@ -106,6 +106,11 @@ func NewAbsenceProof(proofStart, proofEnd int, proofNodes [][]byte, leafHash []b
 	return Proof{proofStart, proofEnd, proofNodes, leafHash, ignoreMaxNamespace}
 }
 
+// IsOfEmptyProof checks whether the proof corresponds to an empty proof as defined in NMT specifications https://github.com/celestiaorg/nmt/blob/master/docs/spec/nmt.md.
+func (proof Proof) IsOfEmptyProof() bool {
```
Comment (cmwaters) [inline proof.go:110]:
```suggestion
// IsEmptyProof checks whether the proof corresponds to an empty proof as defined in NMT specifications https://github.com/celestiaorg/nmt/blob/master/docs/spec/nmt.md.
func (proof Proof) IsEmptyProof() bool {
```
nit

**Review by cmwaters — APPROVED** [review]
👍

File: proof.go (hunk 3 above)
Comment (staheri14) [reply proof.go:110]: Thank you for your suggestion. The reason behind this naming is to maintain consistency with the [IsOfAbsence()](https://github.com/celestiaorg/nmt/blob/0c70c1bde94bf2686e854a072715eaf361ac59ef/proof.go#L68) function, that implements a similar logic.

Comment (staheri14) [reply proof.go]: Oh, I see your point, renamed it in https://github.com/celestiaorg/nmt/pull/180/commits/9e3d8de24adb44c98c3ca644b56da63e1f38a715

**Review by rootulp — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**